### PR TITLE
User List

### PR DESCRIPTION
**Pull request by do-kevin**
- ~Forgot to implement `react-table`, will do it later~

### CODE REVIEW

File: src/Screens/Dashboard.js [hunk 1]
```diff
@@ -1,36 +1,140 @@
...
+    );
+  }
+}
+
+const componentWithStyles = withStyles(styles)(Dashboard);
+
+function mapStateToProps(state) {
+  const { profile, userList } = state;
+  return {
+    userProfile: profile,
```
Comment (lan-nguyen91) [inline src/Screens/Dashboard.js]: you need this profile in the listview ?

File: src/Screens/Home.js [hunk 2]
```diff
@@ -106,6 +110,7 @@ class Home extends Component {
                   onClick={() => {
                     logout(history.push("/"));
```
Comment (lan-nguyen91) [inline src/Screens/Home.js]: can you reorder this, redirect should happen at the end, you don't want to redirect and then have function call after, that will cause problems, you are mingling old functionality of old page and new page.

File: src/Screens/Dashboard.js (hunk 1 above)
Comment (do-kevin) [reply src/Screens/Dashboard.js]: What lines?

File: src/redux/state/list/Actions.js [hunk 3]
```diff
@@ -0,0 +1,8 @@
+import { GRAB_USER_LIST, CLEAR_USER_LIST } from "./Constants";
+
+export const grabUserList = (userList = []) => dispatch => dispatch({
+  type: GRAB_USER_LIST,
```
Comment (lan-nguyen91) [inline src/redux/state/list/Actions.js]: this is a very bad naming convention, GET_USER_LIST, or RETRIEVE_USER_LIST

File: src/screens/Dashboard.js [hunk 4]
```diff
@@ -0,0 +1,368 @@
...
+
+  async componentDidMount() {
+    this.refreshList(this.props);
+  }
+
+  refreshList = async (props) => {
+    const { setUserList } = props;
+    await clearUserList();
+    const response = await listAllProfiles();
+    setUserList(response);
```
Comment (lan-nguyen91) [inline src/screens/Dashboard.js]: what is the point of using redux if you gonna use response this way ?

File: src/screens/Dashboard.js [hunk 5]
```diff
@@ -0,0 +1,368 @@
...
+    response.map((result) => {
+      const { user } = result;
+      Object.assign(this.state, { [`modal${user}`]: false });
```
Comment (lan-nguyen91) [inline src/screens/Dashboard.js]: assign state this way is a `no no`

File: src/screens/Dashboard.js [hunk 6]
```diff
@@ -0,0 +1,368 @@
...
+      return false;
+    });
+  }
+
+  openModal = (userId) => {
+    this.setState({
+      [`modal${userId}`]: true,
```
Comment (lan-nguyen91) [inline src/screens/Dashboard.js]: why user_id attach to modal ?

File: src/screens/Dashboard.js [hunk 7]
```diff
@@ -0,0 +1,368 @@
...
+            gridArea: "nav",
+          }}
+        />
+        <AppBar
+          className={classes.topbar}
+        >
+          <Toolbar className={classes.toolbar}>
+            <Button
+              onClick={() => {
+                this.refreshList(this.props);
```
Comment (lan-nguyen91) [inline src/screens/Dashboard.js]: this.props is accessible everywhere inside the component

File: src/services/Api.js [hunk 8]
```diff
@@ -46,8 +47,7 @@ class Api {
       return response.data;
     } catch (error) {
       console.error(error.request);
-      toast.error(error.message);
-      toast.error(error.request.responseText);
+      toast.error(`${error.message}\n${error.request.statusText}`);
```
Comment (lan-nguyen91) [inline src/services/Api.js]: error.request.statusText might not be available all the time, error.request can be null or undefined

File: src/components/ModalCard.js [hunk 9]
```diff
@@ -0,0 +1,54 @@
...
+import {
+  Typography, withStyles, Modal, Slide, Card, CardMedia, Avatar, CardContent,
+} from "@material-ui/core";
+import { bigStyles } from "screens/DashboardStyles";
+
+function ModalCard(props) {
+  const {
+    classes, modalData, openModal, onClose,
+  } = props;
+  const { birth_date, phone_number } = modalData[1];
```
Comment (lan-nguyen91) [inline src/components/ModalCard.js:11]: what is [1] seem like a magic number here

File: src/components/ModalCard.js [hunk 10]
```diff
@@ -0,0 +1,54 @@
...
+              className={classes.media}
+            >
+              <Avatar
+                className={classes.innerAvatar}
+                src={`https://randomuser.me/api/portraits/men/${Math.floor(Math.random() * 100)}.jpg`}
+                alt="avatar placeholder"
+              />
+            </CardMedia>
+            <CardContent>
+              <Typography variant="h5">{modalData[0]}</Typography>
```
Comment (lan-nguyen91) [inline src/components/ModalCard.js:40]: what is 0 ?

File: src/components/buttons/SidebarBtn.js [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+function SidebarBtn(props) {
+  const {
+    classes, label, icon, divider, color, onClick, disabled, to,
+  } = props;
+
+  SidebarBtn.defaultProps = {
+    label: "",
+  };
+
+  SidebarBtn.propTypes = {
```
Comment (lan-nguyen91) [inline src/components/buttons/SidebarBtn.js]: move propTypes outside of function ?

File: src/components/forms/ProfileCreationForm.js [hunk 12]
```diff
@@ -39,11 +38,11 @@ function ProfileCreationForm(props) {
               fname, lname, birthdate, phonenum,
             } = values;
 
-            await initCreation(userId, fname, lname, birthdate, phonenum);
+            await updateProfile(userId, fname, lname, birthdate, phonenum);
```
Comment (lan-nguyen91) [inline src/components/forms/ProfileCreationForm.js]: can you have asycn api call inside action, and don't manually set this, what if the api call failed, then you have stale data in ur redux store ?

File: src/components/forms/ProfileUpdateForm.js [hunk 13]
```diff
@@ -135,13 +134,13 @@ const componentWithStyles = withStyles(styles)(ProfileUpdateForm);
 
 function mapStateToProps(state) {
```
Comment (lan-nguyen91) [inline src/components/forms/ProfileUpdateForm.js]: short hand is probably cleaner

let mapStateToProps = (state) => ({ profile: state.profile });

File: src/screens/Dashboard.js [hunk 14]
```diff
@@ -0,0 +1,198 @@
...
+class Dashboard extends Component {
+  constructor(props) {
+    super(props);
+    this.reactTable = React.createRef();
+  }
+
+  state = defaultState;
+
+  async componentDidMount() {
+    this.refreshList();
```
Comment (lan-nguyen91) [inline src/screens/Dashboard.js:38]: this is fine 👍

File: src/screens/Dashboard.js [hunk 15]
```diff
@@ -0,0 +1,198 @@
...
+  }
+
+  shouldComponentUpdate(nextProps, prevState) {
+    if (prevState.openModal !== this.state.openModal) {
```
Comment (lan-nguyen91) [inline src/screens/Dashboard.js]: this is probably wrong, so only openModal change then we rerender ?
this is naive checking.

File: src/screens/Dashboard.js [hunk 16]
```diff
@@ -0,0 +1,198 @@
...
+          </Toolbar>
+        </AppBar>
+        <main className={classes.rtWrapper}>
+          <ModalCard
+            modalData={this.state.selectedUser}
+            openModal={this.state.openModal}
+            onClose={() => {
+              this.setState({ openModal: false });
+              setTimeout(() => {
+                this.setState(defaultState.selectedUser);
```
Comment (lan-nguyen91) [inline src/screens/Dashboard.js]: i don't think you need to clean the data here, and even if you do, why not using the above setState, why need setTimeout

File: src/components/ModalCard.js (hunk 9 above)
Comment (do-kevin) [reply src/components/ModalCard.js:11]: It's
```
const userData = [
  props.value, // Full name
  props.original, // other data
 ];
```
which is set to state's `selectedUser`:
```
{
  selectedUser: [
    "",
    {
      birth_date: "",
      phone_number: "",
    },
  ],
  openModal: false,
}
```
Then I pass `this.state.selectedUser` into `<ModalCard/>`'s `modalData` prop.

File: src/components/ModalCard.js (hunk 10 above)
Comment (do-kevin) [reply src/components/ModalCard.js:40]: It's `props.value` from line 94 in `Dashboard.js`

File: src/components/ModalCard.js (hunk 9 above)
Comment (lan-nguyen91) [reply src/components/ModalCard.js:11]: why is selectedUser have to be an array ?

Comment (lan-nguyen91) [reply src/components/ModalCard.js:11]: can you not call this modalcard, so generic, if this is profile, name it ProfileCard

**Review by lan-nguyen91 — APPROVED** [review]

**Pull request merged**